### PR TITLE
Add copy button to structured logs message column

### LINKED ISSUE #2109: Add copy button to structured logs message column
A copy button should be available to copy the message. Do the same thing as the source column on the resources grid.

### PR DESCRIPTION
**Pull request by adamint**
Resolves #2109 

`word-wrap: nowrap` had to be added to GridValue's css as otherwise the content could overflow to another line; this property is beneficial for the other uses of GridValue too, which can suffer from the same problem.

image

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2344)

### CODE REVIEW

**Review by JamesNK — CHANGES_REQUESTED** [review]
image

1. The background color of the copy button doesn't look good. Should it be transparent when not hovered over? That's the same experience as non-error/warning rows, it just so happens the background colors match.
2. The exception details icon is on a different line.

**Comment by adamint** [comment]
Fixed both.
image

**Comment by JamesNK** [comment]
image

1. The button hover color has also disappeared. I don't have a strong opinion about whether a hover color makes sense when hovering on a row with a color, but it's now been removed for all copy buttons.
2. The copy button is no longer at the end of the column.

Compare with resources page:

image

**Comment by adamint** [comment]
> The button hover color has also disappeared. I don't have a strong opinion about whether a hover color makes sense when hovering on a row with a color, but it's now been removed for all copy buttons.

I made it so the background is only transparent on error.

> The copy button is no longer at the end of the column.

I do not think the copy button should be at the end of the column. Considering how long the message column can be, it is unlikely to be discoverable.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor [hunk 1]
```diff
@@ -8,47 +8,59 @@
...
-        </div>
-    </FluentTooltip>
-}
+        </FluentTooltip>
 
-<FluentHighlighter HighlightedText="@FilterText" Text="@LogEntry.Message" />
+        <GridValue Value="@LogEntry.Message"
+                   EnableHighlighting="true"
+                   HighlightText="@FilterText"
+                   BackgroundColor="transparent" />
```
Comment (tlmii) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor:55]: Applying BackgroundColor to a FluentButton (Which is what this has the effect of doing) puts the FluentButton into sort of an odd mode where the styling is detached from the normal fluent styling. In fact, I just noticed there's a bug where it generates invalid inline style blocks too (I'll file that shortly with the library).

I think there's an easier way to fix the complaint that James had about losing the hover color:

1. Back out the changes in the most recent commit (https://github.com/dotnet/aspire/pull/2344/commits/4f254e4c60932df644b6af8f0a907a2726a05b82)
2. Remove the two `BackgroundColor="transparent"` attributes from GridValue.razor that were added in https://github.com/dotnet/aspire/pull/2344/commits/7c065ed4b709f3c1349c2a27e00d196be288ca87
3. In StructuredLogs, change these lines:

```css
::deep .log-row-critical {
    background-color: var(--log-critical);
}

::deep .log-row-error {
    background-color: var(--log-error);
}

::deep .log-row-warning {
    background-color: var(--log-warning);
}
```

to this:

```css
::deep .log-row-critical,
::deep .log-row-critical fluent-button[appearance=lightweight]:not(:hover)::part(control) {
    background-color: var(--log-critical);
}

::deep .log-row-error,
::deep .log-row-error fluent-button[appearance=lightweight]:not(:hover)::part(control) {
    background-color: var(--log-error);
}

::deep .log-row-warning,
::deep .log-row-warning fluent-button[appearance=lightweight]:not(:hover)::part(control){
    background-color: var(--log-warning);
}
```

That has the side effect of making the View button in the last column also "transparent":

image

Hover colors on the copy button and the View button are then consistent as well (they were different in the current commit of this PR because of the opacity mode that the FluentButton got into).

File: src/Aspire.Dashboard/Components/Controls/GridValue.razor [hunk 2]
```diff
@@ -27,20 +27,26 @@
...
+                  @attributes="ButtonAttributes">
         <FluentIcon Class="copy-icon" Style="display:inline;" Icon="Icons.Regular.Size16.Copy" />
         <FluentIcon Class="checkmark-icon" Style="display:none;" Icon="Icons.Regular.Size16.Checkmark" />
     </FluentButton>
     <FluentTooltip Anchor="@_anchorId" Position="TooltipPosition.Top">@PreCopyToolTip</FluentTooltip>
 </div>
+
+@code {
+    [Parameter(CaptureUnmatchedValues = true)]
+    public Dictionary<string, object>? ButtonAttributes { get; set; }
```
[the remark this answers is not in the document]
Comment (adamint) [reply src/Aspire.Dashboard/Components/Controls/GridValue.razor:51]: removed the splatting

**Review by tlmii — APPROVED** [review]
Much cleaner! Thanks!

I do think the "should the copy button be all the way at the end" is worth a conversation before GA but relative to adding the copy button functionality, I don't think its worth blocking the PR at the moment.

**Review by JamesNK — CHANGES_REQUESTED** [review]
> I do not think the copy button should be at the end of the column. Considering how long the message column can be, it is unlikely to be discoverable. Contrast this with the resources page, which has a small source column width

The source column displays it at the end of the column. Also, the property grid has the button at the end of the column, which can be long if displayed horizontally.

For example:
copy-button

This isn't urgent for P4 and doesn't need to be rushed in. Please make it consistent.

If we want to change copy buttons in columns across the app then that should be separate issue to consider properly.

**Comment by joperezr** [comment]
Friendly reminder that this PR is assigned to the Preview 4 milestone which is snapping in a couple of hours. Please ensure it is merged by 4pm PST in order to ensure the change will be part of Preview 4.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor (hunk 1 above)
Comment (tlmii) [reply src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor]: > (I'll file that shortly with the library).

Filed https://github.com/microsoft/fluentui-blazor/issues/1598

**Comment by adamint** [comment]
@JamesNK updated
image

I also added disposal of the fluent tooltip when the grid value is disposed

**Comment by adamint** [comment]
No change to existing buttons
image

File: src/Aspire.Dashboard/Components/Controls/GridValue.razor [hunk 3]
```diff
@@ -26,7 +26,7 @@
     }
     @if (EnableMasking)
     {
-        <FluentButton Appearance="Appearance.Lightweight"
+        <FluentButton Appearance="Appearance.Lightweight"y:
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/GridValue.razor:29]: what is this for? Looks like a mistake

File: src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs [hunk 4]
```diff
@@ -82,4 +84,9 @@ private string TrimLength(string? text)
 
         return text ?? "";
     }
+
+    public void Dispose()
+    {
+        _tooltipComponent?.Dispose();
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs:90]: Why is this necessary? There are other places we're using tooltip and not disposing.

**Comment by JamesNK** [comment]
The cell browser tooltip shows up:

image

I added changes to the error icon to fix this happening. Add them to the copy button.

File: src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs (hunk 4 above)
Comment (adamint) [reply src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs:90]: This will close the tooltip. I could only reproduce the issue you filed about the tooltips not closing on data grid data update on tooltips located in the data grid itself. So I don’t think it’s necessary elsewhere

Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs:90]: Which issue is that? I'd like to test it works.

There should be a comment saying why it's being disposed.

Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs:90]: This one: https://github.com/dotnet/aspire/issues/2704

Test on the stress project. The resources that update have this behavior.

**Comment by adamint** [comment]
Hidden now

File: src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs (hunk 4 above)
Comment (adamint) [reply src/Aspire.Dashboard/Components/Controls/GridValue.razor.cs:90]: I am unable to run the stress project. Starting some of the projects fails

**Review by JamesNK — APPROVED** [review]

**Review by JamesNK — APPROVED** [review]
